ext/KrylovKitChainRulesCoreExt/svdsolve.jl

```diff
@@ -0,0 +1,289 @@
+    end
+    UdΔAV = (aUdΔU .+ aVdΔV) .* safe_inv.(vals' .- vals, tol) .+
+            (aUdΔU .- aVdΔV) .* safe_inv.(vals' .+ vals, tol)
+    if !(Δvals isa ZeroTangent)
+        UdΔAV[diagind(UdΔAV)] .+= real.(Δvals)
+    end
+
+    xs = similar(lvecs, n)
+    ys = similar(rvecs, n)
+    for i in 1:n
+        x = scale(lvecs[1], UdΔAV[1, i] / 2)
+        y = scale(rvecs[1], conj(UdΔAV[i, 1]) / 2)
+        for j in 2:n
+            x = VectorInterface.add!!(x, lvecs[j], UdΔAV[j, i] / 2)
+            y = VectorInterface.add!!(y, rvecs[j], conj(UdΔAV[i, j]) / 2)
+        end
+        xs[i] = x
+        ys[i] = y
+    end
+
+    sylvesterargx = similar(lvecs)
+    for i in 1:n
+        x = zerovector(lvecs[1])
+        if !(Δlvecs[i] isa AbstractZero)
+            x = VectorInterface.add!!(x, Δlvecs[i], +1)
+            for j in 1:n
+                y = VectorInterface.add!!(x, lvecs[j], -UdΔU[j, i])
+            end
+        end
+        sylvesterargx[i] = x
+    end
+    sylvesterargy = similar(rvecs)
+    for i in 1:n
+        y = zerovector(rvecs[1])
+        if !(Δrvecs[i] isa AbstractZero)
+            y = VectorInterface.add!!(y, Δrvecs[i], +1)
+            for j in 1:n
+                y = VectorInterface.add!!(y, rvecs[j], -VdΔV[j, i])
+            end
+        end
+        sylvesterargy[i] = y
+    end
+
+    W₀ = (zerovector(lvecs[1]), zerovector(rvecs[1]), fill(one(T), n))
+    QU = orthogonalcomplementprojector(lvecs, n)
+    QV = orthogonalcomplementprojector(rvecs, n)
+    rvals, Ws, reverse_info = let QU = QU, QV = QV, ΔU = sylvesterargx, ΔV = sylvesterargy
+        eigsolve(W₀, n, :LR, alg_rrule) do w
+            x, y, z = w
+            x′ = QU(apply_normal(f, y))
+            y′ = QV(apply_adjoint(f, x))
+            @inbounds for i in 1:length(z)
+                x′ = VectorInterface.add!!(x′, ΔU[i], -z[i])
+                y′ = VectorInterface.add!!(y′, ΔV[i], -z[i])
+            end
+            return (x′, y′, vals .* z)
+        end
+    end
+    if info.converged >= n && reverse_info.converged < n && alg_rrule.verbosity >= 0
+        @warn "`svdsolve` cotangent problem did not converge, whereas the primal singular value problem did"
+    end
+
+    # cleanup and construct final result
+    tol = alg_rrule.tol
+    for i in 1:n
+        x, y, z = Ws[i]
+        _, ic = findmax(abs, z)
+        if ic != i
+            @warn "`svdsolve` cotangent linear problem ($ic) returns unexpected result"
+        end
+        factor = 1 / z[ic]
+        z[ic] = zero(z[ic])
+        error = max(norm(z, Inf), abs(rvals[i] - vals[ic]))
+        if error > 5 * tol && alg_rrule.verbosity >= 0
+            @warn "`svdsolve` cotangent linear problem ($ic) returns unexpected result: error = $error"
+        end
+        xs[ic] = VectorInterface.add!!(xs[ic], x, -factor)
+        ys[ic] = VectorInterface.add!!(ys[ic], y, -factor)
+    end
+    return xs, ys
+end
+
```

Comment:
It might be possible to merge some of the underlying methods by explicitly using `rrule_via_ad(config, apply_adjoint, ...)` and the `apply_normal` variants. 
Similarly, it might make sense to define `rrule`s for these functions, as we have access to both normal and adjoint function applications? This helps with the compile times and can be a lot more convenient in allowing mutable steps within the function, even when the total function is non-mutating. (e.g. preallocating an array and then filling it)